subs/pantry/src/Pantry/Storage.hs

```diff
@@ -436,30 +448,83 @@ loadHackageTarballInfo name version = do
   where
     go (Entity _ ht) = (hackageTarballSha ht, hackageTarballSize ht)
 
+getKey dat = case dat of
+               Left (Entity key _) -> key
+               Right key -> key
+
+storeCabalFile :: (HasPantryConfig env, HasLogFunc env, HasProcessContext env) => ByteString -> P.PackageName -> TreeId -> P.TreeEntry -> ReaderT SqlBackend (RIO env) BlobId
+storeCabalFile cabalBS pkgName tid tentry = do
+  (bid, _) <- storeBlob cabalBS
+  let cabalFile = P.cabalFileName pkgName
+  fid <- insertBy FilePath { filePathPath = cabalFile }
+  return bid
+
+generateHPack :: (HasPantryConfig env, HasLogFunc env, HasProcessContext env) => TreeId -> P.TreeEntry -> VersionId -> ReaderT SqlBackend (RIO env) ()
+generateHPack tid tree@(P.TreeEntry (P.BlobKey hpackSha _) _) vid = do
+  hpackId <- loadBlobBySHA hpackSha
+  hpack' <- case hpackId of
+             Just hpack -> pure hpack
+             Nothing -> error $ "generateHPack:  BlobKey not found: " ++ show (tree, hpackSha)
+  hpackBS <- loadBlobById hpack'
+  (pkgName, cabalBS) <- lift $ hpackToCabal hpackBS
+  bid <- storeCabalFile cabalBS pkgName tid tree
+  let cabalFile = P.cabalFileName pkgName
+  fid <- insertBy FilePath { filePathPath = cabalFile }
+  let hpackRecord = HPack {
+                      hPackTree = tid,
+                      hPackVersion = vid,
+                      hPackCabal = bid,
+                      hPackPath = getKey fid
+                    }
+  insertBy hpackRecord
+  return ()
+
+hpackVersionId :: (HasPantryConfig env, HasLogFunc env, HasProcessContext env) => ReaderT SqlBackend (RIO env) VersionId
+hpackVersionId = do
+  hpackSoftwareVersion <- lift $ hpackVersion
+  hid <- insertBy $ Version { versionVersion = P.VersionP hpackSoftwareVersion }
+  case hid of
+    Left (Entity versionId _) -> return versionId
+    Right key -> return key
```

Comment:
Aren't these lines almost the same as in the function `getKey` above?

Reply:
Fixed, thanks.